### PR TITLE
SSO failover

### PR DESCRIPTION
**Pull request by it33**

### CODE REVIEW

**Comment by it33** [comment]
Have gotten repeated questions about SSO failover, and solutions 1 and 2 are really what people should do and option 3 is basically "if you didn't do 1 or 2, this is how you could recover from something catastrophic".

File: source/administration/backup.rst [hunk 1]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+a) Users who's session token expires during the outage
+
+b) Users trying to log in to new devices 
+
+In each case, the user cannot reach the SSO provider, and cannot log in. In this case, there are several potential mitigations: 
+
+1) Configure your SSO provider for high availability 
+
+If you're using a self-hosted single-sign-on provider like Active Directory, several options are available for `high availability configurations that protect your system from unplanned outages. <https://docs.microsoft.com/en-us/microsoft-identity-manager/pam/high-availability-disaster-recovery-considerations-bastion-environment>`_
```
Comment (enahum) [inline source/administration/backup.rst]: I don't think AD or LDAP is a kind of SSO, maybe refer here to OAuth 2.0 or SAML as SSO's

File: source/administration/backup.rst [hunk 2]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
 
 A "complete" disaster recovery solution would protect against both real-time hardware failures using high availability, data corruption failures using automated, and failures of the primary data center by offering both offsite backup and offsite redundant infrastructure. Because the complexity of a full disaster recovery solution is high, it is common for customers to consider trade-offs in cost and complexity relative to the anticipated risks and target recovery times.
 
+Failover from single-sign-in outage
```
Comment (lfbrock) [inline source/administration/backup.rst]: "single sign-on"

File: source/administration/backup.rst [hunk 3]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
 
 A "complete" disaster recovery solution would protect against both real-time hardware failures using high availability, data corruption failures using automated, and failures of the primary data center by offering both offsite backup and offsite redundant infrastructure. Because the complexity of a full disaster recovery solution is high, it is common for customers to consider trade-offs in cost and complexity relative to the anticipated risks and target recovery times.
 
+Failover from single-sign-in outage 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When using single-sign-on with Mattermost Enterprise Edition an outage to your SSO provider can cause a partial outage on your Mattermost instance.
```
Comment (lfbrock) [inline source/administration/backup.rst]: "single sign-on"

File: source/administration/backup.rst [hunk 4]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+When using single-sign-on with Mattermost Enterprise Edition an outage to your SSO provider can cause a partial outage on your Mattermost instance. 
+
+**What happens during an SSO outage?**
+
+_Most people can still log in_ - By default, when a user logs in to Mattermost they receive a session token lasting 30 days (the duration can be configured in the System Console). During an SSO outage, users with valid session tokens can continue to using Mattermost uninterrupted. 
+
+_Some people can't log in_ - During an SSO outage, there are two situations under which a user cannot log in: 
+
+a) Users who's session token expires during the outage
```
Comment (lfbrock) [inline source/administration/backup.rst]: whose

File: source/administration/backup.rst [hunk 5]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+b) Users trying to log in to new devices 
+
+In each case, the user cannot reach the SSO provider, and cannot log in. In this case, there are several potential mitigations: 
+
+1) Configure your SSO provider for high availability 
+
+If you're using a self-hosted single-sign-on provider, several options are available for `high availability configurations that protect your system from unplanned outages. <https://docs.microsoft.com/en-us/microsoft-identity-manager/pam/high-availability-disaster-recovery-considerations-bastion-environment>`_
```
Comment (lfbrock) [inline source/administration/backup.rst]: single sign-on

File: source/administration/backup.rst [hunk 6]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+For SaaS-based authentication providers, while you still have a dependency on service uptime, you can setup redundacy of source systems from which data is being pulled. For example, with the OneLogin SaaS-based authentication service, you can setup `high availabilty LDAP connectivity <https://support.onelogin.com/hc/en-us/articles/204262680-High-Availability-for-LDAP>`_ to further reduce the chances of an outage. 
+
+2) Setup your own IDP to provide an automated or manual SSO failover option
```
Comment (lfbrock) [inline source/administration/backup.rst]: "Set up"

File: source/administration/backup.rst [hunk 7]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+If you're using a self-hosted single-sign-on provider, several options are available for `high availability configurations that protect your system from unplanned outages. <https://docs.microsoft.com/en-us/microsoft-identity-manager/pam/high-availability-disaster-recovery-considerations-bastion-environment>`_
+
+For SaaS-based authentication providers, while you still have a dependency on service uptime, you can setup redundacy of source systems from which data is being pulled. For example, with the OneLogin SaaS-based authentication service, you can setup `high availabilty LDAP connectivity <https://support.onelogin.com/hc/en-us/articles/204262680-High-Availability-for-LDAP>`_ to further reduce the chances of an outage. 
+
+2) Setup your own IDP to provide an automated or manual SSO failover option 
+
+Create a custom Identity Provider for SAML authentication that connects to both an active and a standby authentication option, that can be manually or automatically switched in case of an outage. 
+
+This configuration, security should be carefully reviewed to avoid the standby SSO option to weakening your authentication protections.
```
Comment (lfbrock) [inline source/administration/backup.rst]: this sentence doesn't grammatically make sense.

File: source/administration/backup.rst (hunk 2 above)
Comment (it33) [inline source/administration/backup.rst]: Fixing...

File: source/administration/backup.rst (hunk 4 above)
Comment (it33) [inline source/administration/backup.rst]: Fixing...

File: source/administration/backup.rst (hunk 5 above)
Comment (it33) [inline source/administration/backup.rst]: Fixing...

File: source/administration/backup.rst [hunk 8]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+b) Users trying to log in to new devices 
+
+In each case, the user cannot reach the SSO provider, and cannot log in. In this case, there are several potential mitigations: 
+
+1) Configure your SSO provider for high availability 
+
+If you're using a self-hosted single-sign-on provider, several options are available for `high availability configurations that protect your system from unplanned outages. <https://docs.microsoft.com/en-us/microsoft-identity-manager/pam/high-availability-disaster-recovery-considerations-bastion-environment>`_
+
+For SaaS-based authentication providers, while you still have a dependency on service uptime, you can setup redundacy of source systems from which data is being pulled. For example, with the OneLogin SaaS-based authentication service, you can setup `high availabilty LDAP connectivity <https://support.onelogin.com/hc/en-us/articles/204262680-High-Availability-for-LDAP>`_ to further reduce the chances of an outage.
```
Comment (it33) [inline source/administration/backup.rst]: Fixing...

File: source/administration/backup.rst (hunk 6 above)
Comment (it33) [inline source/administration/backup.rst]: Fixing...

File: source/administration/backup.rst (hunk 7 above)
Comment (it33) [inline source/administration/backup.rst]: Fixing...

File: source/administration/backup.rst [hunk 9]
```diff
@@ -59,3 +59,52 @@ A properly deployed high availability setup automatically switches over to a red
...
+
+For SaaS-based authentication providers, while you still have a dependency on service uptime, you can setup redundacy of source systems from which data is being pulled. For example, with the OneLogin SaaS-based authentication service, you can setup `high availabilty LDAP connectivity <https://support.onelogin.com/hc/en-us/articles/204262680-High-Availability-for-LDAP>`_ to further reduce the chances of an outage. 
+
+2) Setup your own IDP to provide an automated or manual SSO failover option 
+
+Create a custom Identity Provider for SAML authentication that connects to both an active and a standby authentication option, that can be manually or automatically switched in case of an outage. 
+
+This configuration, security should be carefully reviewed to avoid the standby SSO option to weakening your authentication protections. 
+
+3) Setup a manual failover plan for SSO outages
```
Comment (it33) [inline source/administration/backup.rst]: Fixing...

**Comment by it33** [comment]
@lfbrock wow, really great review, thank you! :) 

giphy

**Comment by esethna** [comment]
+1 after fixing minor typo

**Pull request merged**